crates/volta-core/src/run/binary.rs

```diff
@@ -20,6 +20,17 @@ pub(super) fn command(exe: &OsStr, args: &[OsString], session: &mut Session) ->
     if let Some(project) = session.project()? {
         // Check if the executable is a direct dependency
         if project.has_direct_bin(exe)? {
+            if project
```

Comment:
Since this is on the hot path and will require some extra file stat checks, would it be possible to make this lazy with the below `find_bin` call? That call will return `None` if the expected `node_modules/.bin/my_binary` doesn't exist, so we can check _then_ for the PnP behavior. If it's not around, we can error out (in the same way that we do now). That way we only check for the `.pnp.js` file iff we know the binaries aren't already in the `.bin` folder.

Thinking something like:

```rust
match project.find_bin(exe) {
    Some(path_to_bin) => {
        // Existing code for a project binary
    }
    None => {
        if project.platform().map_or(false, |platform| platform.yarn.is_some()) && project.is_plug_n_play() {
            // This block executing `yarn <command>`
        } else {
            // No binary _and_ no PnP
            return Err(ErrorKind::ProjectLocalBinaryNotFound {
                command: exe.to_string_lossy().to_string(),
            }
            .into());
        }
    }
}
```

Reply:
Makes sense - I updated that to work like how you suggested above